### PR TITLE
WT-10967 Rewrite the "wiredtiger memory model" comment in gcc.h

### PR DESCRIPTION
**Pull request by luke-pearson**
This comment aims to describe wiredtiger's memory model. It is an adaption and additional to what Chenhao wrote in a PM-3220 document in the past.

This PR isn't urgent and is a lot to review. So please take your time reviewing it.

### CODE REVIEW

File: src/include/gcc.h [hunk 1]
```diff
@@ -27,71 +27,145 @@
...
- * an adjacent memory location.
+ * ## Hardware Requirements ##
+ * # Atomic reads and writes #
+ * WiredTiger requires that pointers (void *) and variables that are 8, 16, 32 and 64 bits to be
+ * read/written to atomically. For example if 8-bits of a 32-bit memory location were written and
+ * then the rest of the 32-bits were written, and another thread of control was able to read the
+ * memory location after the first 8-bits were written and before the subsequent 24-bits were
+ * written, WiredTiger would read invalid states. If two threads of control attempt to write to the
+ * same location simultaneously, the result must be one of the two values, not some combination of
+ * both.
```
Comment (quchenhao) [inline src/include/gcc.h:44]: the both

File: src/include/gcc.h [hunk 2]
```diff
@@ -27,71 +27,145 @@
...
- * values across a barrier.
+ * ## Managing concurrency ##
+ * # Volatile #
+ * A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+ * compiler optimization. Leveraging the definitions from Is Parallel Programming Hard, And, If So,
+ * What Can You Do About It?, Paul E. McKenney, WiredTiger utilizes volatile to prevent the
+ * following type of compiler optimizations: Load fusing, store fusing, invented loads, invented
+ * stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+ * WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+ * optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
```
Comment (quchenhao) [inline src/include/gcc.h:76]: Can we explain why we need to define the frequently used concurrent variables as volatile?

File: src/include/gcc.h [hunk 3]
```diff
@@ -27,71 +27,145 @@
...
+ * condition variables.
+ *
+ * It is expected that all the reads and writes of a shared variable protected by a read-write lock
+ * will be performed under the lock. For the same reason as spinlocks, WiredTiger may allow usages
+ * that read the shared variables without taking the lock. Such instances are only permitted if
+ * reading inconsistent data is sufficient for the use case.
+ *
+ * # Atomic operations  #
+ * As previously mentioned WiredTiger already expects loads and stores to specific sizes of points
+ * to be atomic and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
```
Comment (quchenhao) [inline src/include/gcc.h:128]: Should we describe the memory consistency guarantees of these RMW operations?

File: src/include/gcc.h [hunk 4]
```diff
@@ -27,71 +27,145 @@
...
+ *
+ * WiredTiger defines barriers by describing which kind of reordering they prevent, for example a
+ * LoadLoad barrier prevents load instructions before the barrier from being reordered with load
+ * instructions after the barrier. Using this definition we can define 4 types of reorderings,
+ * LoadLoad, StoreStore, LoadStore and StoreLoad. To simplify the barrier semantics further
+ * WiredTiger defines acquire and release barriers, firstly we have the WT_ACQUIRE_READ_WITH_BARRIER
+ * macro which prevents LoadLoad and LoadStore reorderings. Next we have the
+ * WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings, these
+ * barriers should be used as pairs. If one location requires an acquire barrier then there must be
+ * another location that requires a release barrier. Two more barrier constructs are defined in
```
Comment (quchenhao) [inline src/include/gcc.h:150]: If one location requires an acquire barrier then there must be another location that requires a release barrier.

In the opposite direction, it is not true. We can read with relaxed ordering even the data is written with the release barrier. We should make this clear as well.

File: src/include/gcc.h [hunk 5]
```diff
@@ -27,71 +27,145 @@
...
+ *
+ * Previously WiredTiger defined a WT_PUBLISH macro which prevented StoreStore reordering, this
+ * macro has since been removed however the term publish still exists in the code base. This is
+ * used to refer to writing a value to a shared memory location. It doesn't define any memory
+ * ordering semantics.
+ *
+ * # Marked access #
+ * Barriers are a heavy hammer for situations that only need to prevent reordering of instructions
+ * on one side of a load or store. To reduce the amount of reorderings prevented, and to access an
+ * ARM specific optimization WiredTiger defines two marked access macros. These are loads and stores
```
Comment (quchenhao) [inline src/include/gcc.h:163]: Should we mention ARM here? We should make our model as generic as possible.

File: src/include/gcc.h [hunk 6]
```diff
@@ -27,71 +27,145 @@
...
+ * that prevent reordering from one direction across that load or store. Firstly there is the
+ * WT_ACQUIRE_READ, this prevents loads and stores following the load from being reordered across
+ * it. Then we have the WT_RELEASE_WRITE macro which prevents loads and stores prior to the store
+ * from being reordered across it. As with the acquire and release barriers these marked access
+ * macros should be paired together to function correctly.
```
Comment (quchenhao) [inline src/include/gcc.h:168]: They are not necessarily paired.

File: src/include/gcc.h (hunk 4 above)
Comment (ajmorton) [reply src/include/gcc.h:150]: I don't agree with this. Barriers enforce an order of operations between two locations which a relaxed atomic load isn't able to do. Without a barrier we're not controlling for the invalidate queue introducing ordering issues.

In general I also prefer the simple mental model of always pairing barriers. The write barrier enforces `write A then B` and the read barrier enforces `read B then A`, and it's easy to see that when the reader sees `B` it must also see `A`

File: src/include/gcc.h [hunk 7]
```diff
@@ -27,71 +27,145 @@
...
+ * memory locations it can result in those threads seeing what would otherwise be impossible values.
+ *
+ * WiredTiger defines barriers by describing which kind of reordering they prevent, for example a
+ * LoadLoad barrier prevents load instructions before the barrier from being reordered with load
+ * instructions after the barrier. Using this definition we can define 4 types of reorderings,
+ * LoadLoad, StoreStore, LoadStore and StoreLoad. To simplify the barrier semantics further
+ * WiredTiger defines acquire and release barriers, firstly we have the WT_ACQUIRE_READ_WITH_BARRIER
+ * macro which prevents LoadLoad and LoadStore reorderings. Next we have the
+ * WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings, these
+ * barriers should be used as pairs. If one location requires an acquire barrier then there must be
```
Comment (ajmorton) [inline src/include/gcc.h:149]:
```suggestion
 * WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings. These macros are always associated with a read or write to memory respectively. WT_ACQUIRE_READ_WITH_BARRIER ensures that all loads and stores after its associated read must take place after the read, and WT_RELEASE_WRITE_WITH_BARRIER enures that all loads and stores before its associated write take place before the write.
 * These barriers should be used as pairs. If one location requires an acquire barrier then there must be
```
This sentence is a bit messy, feel free to clean up as needed.

File: src/include/gcc.h [hunk 8]
```diff
@@ -27,71 +27,145 @@
...
+ * Previously WiredTiger defined a WT_PUBLISH macro which prevented StoreStore reordering, this
+ * macro has since been removed however the term publish still exists in the code base. This is
+ * used to refer to writing a value to a shared memory location. It doesn't define any memory
+ * ordering semantics.
+ *
+ * # Marked access #
+ * Barriers are a heavy hammer for situations that only need to prevent reordering of instructions
+ * on one side of a load or store. To reduce the amount of reorderings prevented, and to access an
+ * ARM specific optimization WiredTiger defines two marked access macros. These are loads and stores
+ * that prevent reordering from one direction across that load or store. Firstly there is the
```
Comment (ajmorton) [inline src/include/gcc.h:164]:
```suggestion
 * that prevent reordering from one direction across that specific load or store, while barriers provide ordering constraints on all operations on all loads and stores on each side of the barrier. Firstly there is the
```

File: src/include/gcc.h
Comment (ajmorton) [inline src/include/gcc.h:1]: This comment is not exclusive to gcc, but I'll leave it to WT-12538 to decide on a new location.

File: src/include/gcc.h (hunk 4 above)
Comment (quchenhao) [reply src/include/gcc.h:150]: It depends on whether the reader needs the stricter ordering. It is not that for every read, we must pair with the release write. For example, we don't need any barrier to read the update chain but we need to do a release write when we add it to the update chain.

**Comment by marcbutler** [comment]
I'd like to suggest this large body of text does not belong in this header file, rather somewhere like {{src/docs/portability.dox}} feels like a more appropriate place. I think it would allow the text to be edited and copy-edited more easily, and a reference to the document can be left in the header if deemed appropriate.
I'll hold off on review until I hear back on this suggestion.

File: src/include/gcc.h (hunk 4 above)
Comment (luke-pearson) [reply src/include/gcc.h:150]: That would be an address dependent load which we probably don't want to talk about?

**Comment by luke-pearson** [comment]
As we're moving this content into the docs this PR is getting replaced, there are still open conversations here some which require changes made to the content. I will try and migrate the relevant bits but if I miss something please let me know.

**Comment by luke-pearson** [comment]
Closing in favour of #10575 10575

File: src/include/gcc.h (hunk 1 above)
Comment (luke-pearson) [reply src/include/gcc.h:44]: Done in new PR in commit: 54cd94b61

File: src/include/gcc.h [hunk 9]
```diff
@@ -27,71 +27,145 @@
...
- * race.  If that can happen, you must increase the size of the memory type to
- * a type guaranteed to be written atomically in a single cycle, without writing
- * an adjacent memory location.
+ * ## Hardware Requirements ##
+ * # Atomic reads and writes #
+ * WiredTiger requires that pointers (void *) and variables that are 8, 16, 32 and 64 bits to be
+ * read/written to atomically. For example if 8-bits of a 32-bit memory location were written and
+ * then the rest of the 32-bits were written, and another thread of control was able to read the
+ * memory location after the first 8-bits were written and before the subsequent 24-bits were
+ * written, WiredTiger would read invalid states. If two threads of control attempt to write to the
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/include/gcc.h:42]: Done in new PR in 63b4af9a4

File: src/include/gcc.h [hunk 10]
```diff
@@ -27,71 +27,145 @@
...
+ * same location simultaneously, the result must be one of the two values, not some combination of
+ * both.
  *
- * WiredTiger additionally requires atomic writes for 64-bit memory locations,
- * and so cannot run on machines with a 32-bit memory bus.
+ * # No Aggregated Writes of Adjacent Memory Locations #
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/include/gcc.h:46]: If it does then w/e.

File: src/include/gcc.h [hunk 11]
```diff
@@ -27,71 +27,145 @@
...
- * the processor, made before the WT_RELEASE_BARRIER call, be visible to all
- * processors in the system before any memory stores by the processor, made
- * after the WT_RELEASE_BARRIER call, are visible to any processor.  The
- * WT_ACQUIRE_BARRIER macro ensures that all loads before the barrier are complete
- * before any loads after the barrier.  The compiler cannot reorder or cache
- * values across a barrier.
+ * ## Managing concurrency ##
+ * # Volatile #
+ * A controversial tool in the C toolkit. WiredTiger utilizes volatile for one reason, managing
+ * compiler optimization. Leveraging the definitions from Is Parallel Programming Hard, And, If So,
```
[the remark this answers is not in the document]
Comment (luke-pearson) [reply src/include/gcc.h:71]: Done in the new PR: [4f13535](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/4f135358bb7fa14a0605dd48dae4d3bce31acbdc)

File: src/include/gcc.h (hunk 2 above)
Comment (luke-pearson) [reply src/include/gcc.h:76]: Moving this conversation to: https://github.com/wiredtiger/wiredtiger/pull/10575#discussion_r1594802348

**Pull request closed without merging**